File: Algorithms/105. Construct Binary Tree from Preorder and Inorder Traversal/BuildTree.java
//Author: Tushar Jaiswal
//Creation Date: 03/11/2018

/*Given preorder and inorder traversal of a tree, construct the binary tree.
Note: You may assume that duplicates do not exist in the tree.

For example, given
preorder = [3,9,20,15,7]
inorder = [9,3,15,20,7]
Return the following binary tree:

    3
   / \
  9  20
    /  \
   15   7*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return buildTreeHelper(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1);
    }
    
    public TreeNode buildTreeHelper(int[] preorder, int pStart, int pEnd, int[] inorder, int iStart, int iEnd)
    {
        if(pStart > pEnd || iEnd < 0)
        { return null; }
        int rootVal = preorder[pStart];
        TreeNode root = new TreeNode(rootVal);
        int rootIndex = 0;
            
        for(int i = iStart; i <= iEnd; i++)
        {
            if(inorder[i] == rootVal)
            {
                rootIndex = i;
                break;
            }
        }
        
        int leftLength = rootIndex - iStart;
        int rightLength = iEnd - rootIndex;
        
        root.left = buildTreeHelper(preorder, pStart + 1, pStart + leftLength, inorder, iStart, rootIndex - 1);
        root.right = buildTreeHelper(preorder, pEnd - (rightLength - 1), pEnd, inorder, rootIndex + 1, iEnd);
        
        return root;
    }
}

File: Algorithms/106. Construct Binary Tree from Inorder and Postorder Traversal/BuildTree.cs
//Author: Tushar Jaiswal
//Creation Date: 03/10/2018

/*Given inorder and postorder traversal of a tree, construct the binary tree.
Note: You may assume that duplicates do not exist in the tree.

For example, given
inorder = [9,3,15,20,7]
postorder = [9,15,7,20,3]
Return the following binary tree:

    3
   / \
  9  20
    /  \
   15   7*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode BuildTree(int[] inorder, int[] postorder) {
        int length =  postorder.Length;
        if (length == 0)
        { return null; }
        
        int rootVal = postorder[length - 1];
        TreeNode root = new TreeNode(rootVal);
        
        int leftLength = 0, rightLength = 0;
        for(int i = 0; i < length; i++)
        {
            if(inorder[i] == rootVal)
            {
                leftLength = i;
                rightLength = length - (i + 1);
            }
        }
        
        int[] rightInorder = new int[rightLength];
        int[] rightPostorder = new int[rightLength];
        int[] leftInorder = new int[leftLength];
        int[] leftPostorder = new int[leftLength];
        for(int i = 0; i < leftLength || i < rightLength; i++)
        {
            if(i < leftLength)
            {
                leftInorder[i] = inorder[i];
                leftPostorder[i] = postorder[i];
            }
            if(i < rightLength)
            {
                rightInorder[i] = inorder[leftLength + 1 + i];
                rightPostorder[i] = postorder[leftLength + i];
            }
        }
        
        root.right = BuildTree(rightInorder, rightPostorder);
        root.left = BuildTree(leftInorder, leftPostorder);
        
        return root;
    }
}